hoomd/ParticleGroup.cc

```diff
@@ -789,6 +792,100 @@ unsigned int ParticleGroup::intersectionSize(std::shared_ptr<ParticleGroup> othe
     return n;
     }
 
+void ParticleGroup::thermalizeParticleMomenta(Scalar kT, unsigned int seed, unsigned int timestep)
+    {
+    unsigned int group_size = this->getNumMembers();
+
+    const unsigned int D = Scalar(m_sysdef->getNDimensions());
+
+    ArrayHandle<Scalar4> h_vel(m_pdata->getVelocities(),
+                               access_location::host,
+                               access_mode::readwrite);
+
+    ArrayHandle<Scalar4> h_orientation(m_pdata->getOrientationArray(),
+                                       access_location::host,
+                                       access_mode::read);
+
+    ArrayHandle<Scalar4> h_angmom(m_pdata->getAngularMomentumArray(),
+                                  access_location::host,
+                                  access_mode::readwrite);
+
+    ArrayHandle<Scalar3> h_inertia(m_pdata->getMomentsOfInertiaArray(),
+                                   access_location::host,
+                                   access_mode::read);
+
+    ArrayHandle<unsigned int> h_tag(m_pdata->getTags(),
+                                    access_location::host,
+                                    access_mode::read);
+
+    // Total the system's linear momentum
+    vec3<Scalar> tot_momentum(0, 0, 0);
+
+    // Loop over all particles in the group
+    for (unsigned int group_idx = 0; group_idx < group_size; group_idx++)
+        {
+        unsigned int j = this->getMemberIndex(group_idx);
+        unsigned int ptag = h_tag.data[j];
+
+        // Seed the RNG
+        hoomd::RandomGenerator rng(hoomd::RNGIdentifier::ParticleGroupThermalize, seed, ptag, timestep);
+
+        // Generate a random velocity
+        Scalar mass =  h_vel.data[j].w;
+        Scalar sigma = fast::sqrt(kT / mass);
```

Comment:
I _think_ that we looked at this at some point and the fast sqrt isn't actually doing anything special right now (in theory you could do smarter algorithms than what `slow::sqrt` should do using inverses etc). In case I'm wrong or if that ever changes, though, are there any issues with using the less precise sqrt with respect to the accuracy of the resulting velocity distribution?

Reply:
Same question applies to the angular mementa below.

Reply:
The fast and slow sqrt's have more performance implications on the GPU - they should be identical on the CPU. Still, this code may be copied to the GPU some day and a few precise sqrts per particle will not slow performance significantly in this rarely called method. I replaced them with slow::.